src/model/model.nit

```diff
@@ -1188,6 +1248,287 @@ abstract class MType
 	end
 end
 
+# A type that represents an operation over a set of types.
+abstract class MTypeSet[E: MType]
+	super MType
+
+	# The module of reference.
+	#
+	# Used to populate `model`, find primitive types and internally build new
+	# types with the same operator.
+	#
+	# SEE: `apply_to`
+	var mmodule: MModule
+
+	# The embeded types.
+	#
+	# Must not be empty. Must not be edited.
+	var operands: Set[E]
+
+	# Cache of types according to operand sets.
+	#
+	# SEE: `cache`
+	private var general_cache = new HashMap[Set[MType], MType]
+
+	init do
+		assert operands.not_empty
+		general_cache[operands] = self
+	end
+
+	redef fun model do return mmodule.model
+
+	redef fun location do return operands.first.location
+	# TODO: Return a more accurate location.
+
+	# Apply the same operator over the specified types.
+	#
+	# May optimize the resulting expression and not return an instance of
+	# `SELF`.
+	fun apply_to(operands: Set[MType], mmodule: MModule,
+			anchor: nullable MClassType): MType is abstract
+
+	# Return the cached version of this operation over the specified set of operands.
+	#
+	# Intended to be used internally to cache results of methods like
+	# `as_notnull`, `lookup_fixed`, `resolve_for` and `undecorate`.
+	#
+	# SEE: `apply_to`
+	protected fun cache(operands: Set[MType]): MType
```

Comment:
Also, caching is a trade-off between hash table accesses and the cost of `==` between `MType`s (from what I understand from the rest of the model). I hesitated to add the cache at the level of `MModule` because when intersection will be flattened, there may be a lot of temporary types created just to simplify the algorithms. For example, to compute a flattened version of `A and B and C`, we may create `A and B` first, then discard it. If only we had weak references, the choice would be easy.